File: PointSDK-MinimalIntegrationExample-Objc/ViewController.m
//
//  ViewController.m
//  PointSDK-MinimalIntegrationExample-iOS-ObjC
//
//  Created by Bluedot Innovation
//  Copyright (c) 2016 Bluedot Innovation. All rights reserved.
//

#import "ViewController.h"
@import BDPointSDK;

@interface ViewController ()

@end

@implementation ViewController

// Use a project Id acquired from the Canvas UI.
NSString *projectId = @"YourProjectId";

// Use a Tempo Destination Id from the Canvas UI.
NSString * tempoDestinationId = @"YourTempoDestinationId";


- (IBAction)initializeSDkTouchUpInside {
    //MARK: Initialize with Point SDK
    if([BDLocationManager.instance isInitialized] == NO)
    {
        [BDLocationManager.instance initializeWithProjectId:projectId completion:^(NSError * _Nullable error) {
            if(error != nil){
                NSLog(@"Initialisation Error: %@", error.localizedDescription);
                [self showAlertWithTitle:@"Initialisation Error" message:error.localizedDescription];
                return;
            }
            
            NSLog(@"Initialised successfully with Point sdk");
        }];

    }
}

- (IBAction)resetSDkTouchUpInside {
    if([BDLocationManager.instance isInitialized] == YES)
    {
        [BDLocationManager.instance resetWithCompletion:^(NSError * _Nullable error) {
            if(error != nil){
                NSLog(@"Reset Error: %@", error.localizedDescription);
                [self showAlertWithTitle:@"Reset Error" message:error.localizedDescription];
                return;
            }
            
            NSLog(@"Point SDK Reset successfully");
        }];

    }
}

- (IBAction)startGeotriggeringTouchUpInside {
    [BDLocationManager.instance startGeoTriggeringWithCompletion:^(NSError * _Nullable error) {
        if(error != nil) {
            NSLog(@"Start Geotriggering Error: %@", error.localizedDescription);
            [self showAlertWithTitle:@"Start Geotriggering Error" message:error.localizedDescription];
            return;
        }
        
        NSLog(@"Start Geotriggering successfully");
    }];
}

- (IBAction)stopGeotriggeringTouchUpInside {
    [BDLocationManager.instance stopGeoTriggeringWithCompletion:^(NSError * _Nullable error) {
        if(error != nil) {
            NSLog(@"Stop Geotriggering Error: %@", error.localizedDescription);
            [self showAlertWithTitle:@"Stop Geotriggering Error" message:error.localizedDescription];
            return;
        }
        
        NSLog(@"Stop Geotriggering successfully");
    }];
}

- (IBAction)startTempoTouchUpInside {
    [BDLocationManager.instance startTempoTrackingWithDestinationId:tempoDestinationId completion:^(NSError * _Nullable error) {
        if(error != nil) {
            NSLog(@"Start Tempo Error: %@", error.localizedDescription);
            [self showAlertWithTitle:@"Start Tempo Error" message:error.localizedDescription];
            return;
        }
        
        NSLog(@"Start Tempo successfully");
    }];
}

- (IBAction)stopTempoTouchUpInside {
    [BDLocationManager.instance stopTempoTrackingWithCompletion:^(NSError * _Nullable error) {
        if(error != nil) {
            NSLog(@"Stop Tempo Error: %@", error.localizedDescription);
            [self showAlertWithTitle:@"Stop Tempo Error" message:error.localizedDescription];
            return;
        }
        
        NSLog(@"Stop Tempo successfully");
    }];
}

- (IBAction)openLocationSettingsTouchUpInside {
    
    NSURL * settingsUrl = [[NSURL alloc] initWithString: UIApplicationOpenSettingsURLString];
    
    if(settingsUrl == nil) {
        return;
    }
    
    if([UIApplication.sharedApplication canOpenURL:settingsUrl])
    {
        [UIApplication.sharedApplication openURL:settingsUrl options:@{} completionHandler:^(BOOL success) {
            NSLog(@"Settings Opened: %@", success ? @"Yes" : @"No");
        }];
    }
}

- (void)showAlertWithTitle: (NSString *)title message: (NSString *)message
{
    UIAlertController *alertController = [ UIAlertController alertControllerWithTitle: title
                         message: message
                  preferredStyle: UIAlertControllerStyleAlert ];
            
    UIAlertAction *OK = [ UIAlertAction actionWithTitle: @"OK" style: UIAlertActionStyleCancel handler: nil ];
            
    [ alertController addAction:OK ];
            
    [self presentViewController: alertController animated: YES completion: nil];
}

@end

File: PointSDK-MinimalIntegrationExample-Swift/ViewController.swift
//
//  ViewController.swift
//  PointSDK-MinimalIntegrationExample-iOS-Swift
//
//  Created by Bluedot Innovation
//  Copyright (c) 2016 Bluedot Innovation. All rights reserved.
//
//

import UIKit
import BDPointSDK
import UserNotifications

class ViewController: UIViewController {
    // Use a project Id acquired from the Canvas UI.
    var projectId = "YourProjectId"
    
    // Use a Tempo Destination Id from the Canvas UI.
    var tempoDestinationId = "YourTempoDestinationId"
    
    @IBAction func initializeSDKButtonTouchUpInside(_ sender: Any) async {
        //MARK: Initialize with Point SDK
        guard BDLocationManager.instance()?.isInitialized() == false else {
            return
        }
        
        do {
            try await BDLocationManager.instance().initialize(withProjectId: projectId)
            print( "Initialised successfully with Point sdk" )
            BDLocationManager.instance()?.requestAlwaysAuthorization()
        } catch {
            print("Initialisation Error: \(error.localizedDescription)")
            self.showAlert(title: "Initialisation Error", message: error.localizedDescription)
        }
    }
    
    //MARK: Reset SDK
    @IBAction func resetButtonAction(_ sender: UIButton) async {
        guard BDLocationManager.instance()?.isInitialized() == true else {
            return
        }
        
        do {
            try await BDLocationManager.instance()?.reset()
            print( "Point SDK Reset successfully" )
        }
        catch {
            print("Reset Error \(error.localizedDescription)")
            self.showAlert(title: "Reset Error", message: error.localizedDescription)
        }
    }
    
    //MARK: Start GeoTriggering
    @IBAction func startGeotriggeringTouchUpInside(_ sender: Any) async {
        do {
            try await BDLocationManager.instance()?.startGeoTriggering()
            print( "Start Geotriggering successfully" )
        } catch {
            print("Start Geotriggering Error: \(error.localizedDescription)")
            self.showAlert(title: "Start Geotriggering Error", message: error.localizedDescription)
        }
    }
    
    //MARK: Stop GeoTriggering
    @IBAction func stopGeotriggeringButtonTouchUpInside(_ sender: Any) async {
        do {
            try await BDLocationManager.instance()?.stopGeoTriggering()
            print( "Stop Geotriggering successfully" )
        } catch {
            print("Stop Geotriggering Error \(error.localizedDescription)")
            self.showAlert(title: "Stop Geotriggering Error", message: error.localizedDescription)
        }
    }
    
    //MARK: Start Tempo
    @IBAction func startTempoButtonTouchUpInside(_ sender: Any) async {
        do {
            try await BDLocationManager.instance()?.startTempoTracking(withDestinationId: tempoDestinationId)
            print( "Start Tempo successfully" )
        } catch {
            print("Start Tempo Error: \(error.localizedDescription)")
            self.showAlert(title: "Start Tempo Error", message: error.localizedDescription)
        }
    }
    
    //MARK: Stop Tempo
    @IBAction func stopTempoButtonTouchUpInside(_ sender: Any) async {
        do {
            try await BDLocationManager.instance()?.stopTempoTracking()
            print( "Stop Tempo successfully" )
        } catch {
            print("Stop Tempo Error \(error.localizedDescription)")
            self.showAlert(title: "Stop Tempo Error", message: error.localizedDescription)
        }
    }
    
    //MARK: Open Location settings on device
    @IBAction func openDeviceSettingsButtonTouchUpInside(_ sender: Any) async {
        guard let settingsUrl = URL(string: UIApplication.openSettingsURLString) else {
            return
        }
        
        guard UIApplication.shared.canOpenURL(settingsUrl) else {
            return
        }
        
        let result = await UIApplication.shared.open(settingsUrl, options: [:])
        print("Settings opened: \(result)")
    }
    
    
    private func showAlert(title: String,message: String) {
        //MARK:- Show Alert
        let dismissAction = UIAlertAction(title: "Dismiss", style: .cancel, handler: nil)
        let alertController = UIAlertController(title: title, message: message, style: .alert, actions: dismissAction)
        
        self.present(alertController, animated: true, completion: nil)
    }
}
